buffer/src/main/java/io/netty/buffer/PoolArena.java

```diff
@@ -253,73 +415,224 @@ void reallocate(PooledByteBuf<T> buf, int newCapacity, boolean freeOldMemory) {
         buf.setIndex(readerIndex, writerIndex);
 
         if (freeOldMemory) {
-            free(oldChunk, oldHandle);
+            free(oldChunk, oldHandle, oldMaxLength, buf.cache);
         }
     }
 
-    protected abstract PoolChunk<T> newChunk(int pageSize, int maxOrder, int pageShifts, int chunkSize);
-    protected abstract PoolChunk<T> newUnpooledChunk(int capacity);
-    protected abstract PooledByteBuf<T> newByteBuf(int maxCapacity);
-    protected abstract void memoryCopy(T src, int srcOffset, T dst, int dstOffset, int length);
-    protected abstract void destroyChunk(PoolChunk<T> chunk);
+    /**
+     * Returns the number of thread caches backed by this arena.
+     */
+    public int numThreadCaches() {
+        return numThreadCaches.get();
+    }
 
-    public synchronized String toString() {
-        StringBuilder buf = new StringBuilder();
-        buf.append("Chunk(s) at 0~25%:");
-        buf.append(StringUtil.NEWLINE);
-        buf.append(qInit);
-        buf.append(StringUtil.NEWLINE);
-        buf.append("Chunk(s) at 0~50%:");
-        buf.append(StringUtil.NEWLINE);
-        buf.append(q000);
-        buf.append(StringUtil.NEWLINE);
-        buf.append("Chunk(s) at 25~75%:");
-        buf.append(StringUtil.NEWLINE);
-        buf.append(q025);
-        buf.append(StringUtil.NEWLINE);
-        buf.append("Chunk(s) at 50~100%:");
-        buf.append(StringUtil.NEWLINE);
-        buf.append(q050);
-        buf.append(StringUtil.NEWLINE);
-        buf.append("Chunk(s) at 75~100%:");
-        buf.append(StringUtil.NEWLINE);
-        buf.append(q075);
-        buf.append(StringUtil.NEWLINE);
-        buf.append("Chunk(s) at 100%:");
-        buf.append(StringUtil.NEWLINE);
-        buf.append(q100);
-        buf.append(StringUtil.NEWLINE);
-        buf.append("tiny subpages:");
-        for (int i = 1; i < tinySubpagePools.length; i ++) {
-            PoolSubpage<T> head = tinySubpagePools[i];
+    @Override
+    public int numTinySubpages() {
+        return tinySubpagePools.length;
+    }
+
+    @Override
+    public int numSmallSubpages() {
+        return smallSubpagePools.length;
+    }
+
+    @Override
+    public int numChunkLists() {
+        return chunkListMetrics.size();
+    }
+
+    @Override
+    public List<PoolSubpageMetric> tinySubpages() {
+        return subPageMetricList(tinySubpagePools);
+    }
+
+    @Override
+    public List<PoolSubpageMetric> smallSubpages() {
+        return subPageMetricList(smallSubpagePools);
+    }
+
+    @Override
+    public List<PoolChunkListMetric> chunkLists() {
+        return chunkListMetrics;
+    }
+
+    private static List<PoolSubpageMetric> subPageMetricList(PoolSubpage<?>[] pages) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)